### DISCUSSION TITLE
How can we host DAB is azure webapp

### DISCUSSION THREAD

**Discussion by madhugilla** [Azure/data-api-builder#2213] [Q&A]
Hi,

we already have an existing api, we have hosted this in azure webapp, can we host DAB along with this ?, can you share any articles or samples on how to achieve this ?

Thanks -Nen

**Comment by seantleonard** [comment]
Hi @madhugilla,

Thanks for reaching out. Can you share more details about what your end goal is here? When you say "we have hosted this in azure webapp, can we host DAB along with this?" do you mean that you just want to host your instance of DAB in an Azure App Service instance?

We have a doc on deploying to Azure Container apps here: https://learn.microsoft.com/azure/data-api-builder/tutorial-deploy-container-app-cli

**Reply by madhugilla** [reply]
yes, we want to just host the instance of DAB in azure appservice instance, along with our api.

**Reply by JuanMeeske** [reply]
@seantleonard I want to know this as well, the use case for me is to use Azure app service instead of Static webapp because of the (preview) database connection. The client has a policy not to use preview features in prod. So therefore we need a GA way of doing it. 
I have noticed the Container apps and AKS way. But would like to know the App service way. 

This is what i found in the docs:

https://learn.microsoft.com/en-us/azure/data-api-builder/deployment/hosting-options#azure-app-service

**Reply by seantleonard** [reply]
We have a backlog task to author a doc which describes deploying DAB on Azure Web App w/ Custom Container (Azure App Service). 

I've deployed this in a lab and have been able to get DAB to serve requests with that setup. Here is the Azure Web App I believe you're referencing in the post title: https://learn.microsoft.com/en-us/azure/app-service/configure-custom-container?tabs=debian&pivots=container-linux

You would be able to follow similar container deployment via setting up managed identity, database, storage account to mount dab-config.json, and setup an Azure Web App which pulls the dab container image. 

There are many similarities with the steps outlined in this tutorial:
[Deploy to Azure Container Apps - Data API builder | Microsoft Learn](https://learn.microsoft.com/azure/data-api-builder/deployment/how-to-publish-container-apps?context=%2Fazure%2Fdata-api-builder%2Fcontext%2Fcontext)
 
Azure Web Apps w/Containers (Linux Host) require extra steps on the Azure Web Apps side in order to map a custom file share to the DAB container. This isn't really something we can update from the DAB side because this is a specific step related to Azure Web Apps and everyone's file share names are different. You have to map the `dab-config.json` file share from the linux host to the Docker container by enabling persistent storage for the webapp:

az cli command for that: https://learn.microsoft.com/en-us/azure/app-service/configure-custom-container?tabs=debian&pivots=container-linux#use-persistent-shared-storage
 
This allows any files mounted to Linux host's /home folder to then be mounted to the container
Configure a custom container - Azure App Service | Microsoft Learn
Learn how to configure a custom container in Azure App Service. This article shows the most common configuration tasks.
 
```sh
az webapp config appsettings set --resource-group <group-name> --name <app-name> --settings WEBSITES_ENABLE_APP_SERVICE_STORAGE=true
 ```

You'll also need to enable container logging:
> Please use https://aka.ms/linux-diagnostics to enable logging to see container logs here.

Azure Web App w/ Container Configuration changes take a while to take effect: 

1. Set fileshare to mount to /home/dabconfig
    1. Azure Web App blade
    1. Settings/Configuration/Path mappings
    1. Add Environment variable
1. Azure Web App Blade
    1. Settings/Environment Variables
    1. Name: ConfigFileName
    1. Value: /home/dabconfig/dabconfig.json
1. Ensure port mapping is configured to port 5000
    1. `azure cli: az webapp config appsettings set --resource-group $RESOURCE_GROUP_NAME --name $APP_NAME --settings WEBSITES_PORT=5000`
1. Confirmed that DAB is looking at: via *_default_docker.log file
    - 2024-09-17T19:30:10.815947517Z Loading config file from /home/dabconfig/dab-config.json.

**Reply by JuanMeeske** [reply]
@seantleonard Is there also a way to expose the dab-config.json to the webapp using a appsettings? Or does it need to be a seperate json file?

**Reply by RubenCerna2079** [reply]
Hi @JuanMeeske, there is no way to expose the dab-config.json using appsettings. 
Is there any specific reason why you wanted to do this? Because DAB can support multiple environments in a similar way to appsettings.
https://learn.microsoft.com/en-us/azure/data-api-builder/reference-configuration?tabs=http#environments

**Reply by JuanMeeske** [reply]
> Hi @JuanMeeske, there is no way to expose the dab-config.json using appsettings. 
> 
> Is there any specific reason why you wanted to do this? Because DAB can support multiple environments in a similar way to appsettings.
> 
> https://learn.microsoft.com/en-us/azure/data-api-builder/reference-configuration?tabs=http#environments

This if you could, you wont need the storage account anymore.
App settings would be a nicer way of adding config in if you asked me.

**Comment by JuanMeeske** [comment]
I have put this in a Bicep, So let me know if this works for you:

https://github.com/JuanMeeske/AppServiceForContainers-DAB
